add vira-dropdown to vira package

electrovir opened the pull request:
# Changes

- add `vira-dropdown` and `vira-dropdown-item`
- add element-book pages for each

# How to test

Unfortunately I do not have deploy previous setup for this so you'll have to test locally:

1. clone this repo locally
2. checkout this branch
3. run `npm ci`
4. run `cd packages/vira-book`
5. run `npm start`
6. test out the `vira-dropdown` element in the vira-book frontend

electrovir committed:
add vira-dropdown and vira-dropdown-item

electrovir committed:
add element-book pages for vira-dropdown and vira-dropdown-item

sarangspadalkar commented on cspell.config.cjs:
```diff
@@ -8,5 +8,6 @@ module.exports = {
     words: [
         ...baseConfig.words,
         'observavir',
+        'listbox',
```
Test comment

sarangspadalkar commented on packages/vira-book/src/element-book/entries/dropdown/vira-dropdown-item.element.book.ts:
```diff
@@ -0,0 +1,86 @@
+    customStyle?: CSSResult;
+    customTemplate?: HTMLTemplateResult;
+}> = [
+    {
```
Please add examples: 
- having not enough horizontal space
- being stretched to take up more space than necessary

sarangspadalkar commented on packages/vira-book/src/element-book/entries/dropdown/vira-dropdown-item.element.book.ts:
```diff
@@ -0,0 +1,86 @@
+import {ViraDropdownItem} from 'vira';
+import {dropdownPage} from './vira-dropdown.book';
+
+const examples: ReadonlyArray<{
```
Please add examples: 
- having not enough horizontal space
- being stretched to take up more space than necessary

electrovir replied:
done

sarangspadalkar commented on packages/vira-book/src/element-book/entries/dropdown/vira-dropdown-item.element.book.ts:
```diff
@@ -0,0 +1,86 @@
+import {BookPageControlTypeEnum, defineBookPage, definePageControl} from 'element-book';
```
Is this book deployed somewhere? Netlify or github-pages?

electrovir replied:
GitHub Pages: http://electrovir.github.io/vira/book

Thus there is no deploy preview 😔

sarangspadalkar commented on packages/vira-book/src/element-book/entries/dropdown/vira-dropdown.element.book.ts:
```diff
@@ -0,0 +1,206 @@
+            ],
+        },
+    },
+    // {
```
This is your personal repo, so I dont know what conventions you have enforced. But I would avoid having commented code.

electrovir replied:
good catch!

electrovir replied:
fixed

sarangspadalkar commented on packages/vira-book/src/element-book/entries/dropdown/vira-dropdown-item.element.book.ts:
```diff
@@ -0,0 +1,86 @@
+const examples: ReadonlyArray<{
+    title: string;
+    inputs?: typeof ViraDropdownItem.inputsType;
+    customStyle?: CSSResult;
```
this is not used anywhere.

```suggestion
```

electrovir replied:
Now it is :)

sarangspadalkar commented on packages/vira/src/elements/vira-dropdown.element.ts:
```diff
@@ -0,0 +1,511 @@
+    },
+    /** Default colors meet APCA contrast for small body text: https://www.myndex.com/APCA/ */
+    cssVars: {
+        'vira-dropdown-trigger-background-color': 'white',
```
These vars values are very generic which can be used by other elements as well. So add them to a shared css file which can be shared between multiple elements.

electrovir replied:
Agreed, that is planned for the future in `theme-vir` but it's not ready yet.

sarangspadalkar commented on packages/vira/src/elements/vira-dropdown.element.ts:
```diff
@@ -0,0 +1,511 @@
+
+        ${hostClasses['vira-dropdown-disabled'].selector} {
+            ${viraDisabledStyles}
+
```
remove this space

```suggestion
```

electrovir replied:
done

sarangspadalkar commented on packages/vira/src/elements/vira-dropdown.element.ts:
```diff
@@ -0,0 +1,511 @@
+            }
+        }
+
+        const selectedOptions: ReadonlyArray<Readonly<ViraDropdownOption>> = inputs.selected.length
```
I think this should be moved inside a function and that function should have unit tests written for it.

electrovir replied:
done

sarangspadalkar commented on packages/vira/src/elements/vira-dropdown.element.ts:
```diff
@@ -0,0 +1,511 @@
+import {PartialAndUndefined, isLengthAtLeast, joinWithFinalConjunction} from '@augment-vir/common';
```
please add tests for this element.

electrovir replied:
done

electrovir commented on cspell.config.cjs:
todo: prevent opening multiple dropdowns at once

electrovir replied:
fixed

electrovir commented on cspell.config.cjs:
todo: add placeholder input

electrovir replied:
done

electrovir committed:
address feedback

add more element book cases
close other dropdowns when another dropdown is opened
add placeholder
test vira-dropdown
minor tweaks and fixes

sarangspadalkar commented on packages/vira/src/elements/vira-dropdown.element.ts:
```diff
@@ -0,0 +1,561 @@
+import {ViraIcon} from './vira-icon.element';
+
+function assertUniqueIdProps(options: ReadonlyArray<Readonly<{id: PropertyKey}>>) {
+    const usedIds: Record<PropertyKey, boolean> = {};
```
nit: Generally [Sets](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Set) are used to store unique values.

electrovir replied:
good point

electrovir replied:
fixed

sarangspadalkar commented on packages/vira/src/elements/vira-dropdown.element.ts:
```diff
@@ -0,0 +1,561 @@
+        selectionPrefix: string;
+        isDisabled: boolean;
+        /** For debugging purposes only. Very bad for actual production code use. */
+        z_debug_forceOpenState: boolean;
```
question: what is the thinking process while behing the name of this property? Is there a reason its strats with `z`?

electrovir replied:
Starting with `_` puts it at the top of IntelliSense's auto complete list, unfortunately. `z` puts it at the bottom 😆

sarangspadalkar commented on packages/vira/src/elements/vira-dropdown.element.ts:
```diff
@@ -0,0 +1,561 @@
+        'vira-dropdown-item-hover-text-color': '#000000',
+    },
+    styles: ({hostClasses, cssVars}) => css`
+        :host {
```
this styles object is huge, can it be refactored somehow?

electrovir replied:
Maybe some of the children can be extracted into separate elements, I'll take a closer look.

electrovir replied:
I split out `options-wrapper` into a separate element. That cut down on the file and styles size a bit.

sarangspadalkar commented on packages/vira/src/elements/vira-dropdown.element.ts:
```diff
@@ -0,0 +1,561 @@
+        popUpManager: perInstance(() => new PopUpManager()),
+        navController: undefined as NavController | undefined,
+    },
+    cleanupCallback({state, updateState}) {
```
when is this callback executed?

electrovir replied:
https://www.npmjs.com/package/element-vir#other-callbacks

sarangspadalkar commented on packages/vira/src/elements/vira-dropdown.element.ts:
```diff
@@ -0,0 +1,561 @@
+        updateState({showPopUpResult: undefined});
+        state.popUpManager.destroy();
+    },
+    initCallback({state, updateState, host, inputs, dispatch, events}) {
```
when is this callback executed?

electrovir replied:
`initCallback` is called whenever an element is connected to the DOM. It's triggered by the browser itself.

electrovir replied:
https://www.npmjs.com/package/element-vir#other-callbacks

sarangspadalkar commented on packages/vira/src/elements/vira-dropdown.element.ts:
```diff
@@ -0,0 +1,561 @@
+import {PartialAndUndefined, joinWithFinalConjunction} from '@augment-vir/common';
```
this file is 500 line long, consider moving these utils functions out of this file.

electrovir replied:
good idea

electrovir replied:
functions moved

electrovir committed:
split functions out of vira-dropdown file, organize dropdown files

electrovir committed:
add form themes

electrovir committed:
split out dropdown options into a dedicated element
